graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,184 @@
+        # UUIDs are uniformly distributed, but it is not wise to make more pages than
+        # there are values.
+        class_count = schema_info.statistics.get_class_count(pagination_node)
+        page_capacity = min(page_capacity, class_count)
+    else:
+        # If the ratio len(quantiles) // number_of_pages is N, then the largest page is
+        # expected to be (N + 1) / N times larger than the smallest page. This is because
+        # if we assume that quantiles divide the domain into equal chunks, some pages will
+        # span over N + 1 chunks, and some will span over N chunks. So if N is 5,
+        # (N + 1) / N = 1.2, which is not too bad.
+        ideal_min_num_quantiles_per_page = 5
+        ideal_quantile_resolution = ideal_min_num_quantiles_per_page * number_of_pages + 1
+
+        quantiles = schema_info.statistics.get_field_quantiles(pagination_node, pagination_field)
+        if quantiles is None:
+            # If there's no quantiles, we don't paginate. We could try to assume an uniform
+            # value distribution and make some pagination plan with more than 1 page, but
+            # the cost estimator would need to match the behavior and assume uniform value
+            # distribution when estimating the selectivity of range filters.
+            page_capacity = 1
+        elif len(quantiles) - 1 < number_of_pages:
+            # If we have some quantiles but not enough, we generate a page for each chunk
+            # of values separated by the quantiles.
+            page_capacity = len(quantiles) - 1
+
+        # If we have less than the ideal number of quantiles, we request for more.
+        if quantiles is None or len(quantiles) < ideal_quantile_resolution:
+            current_quantile_resolution = 0 if quantiles is None else len(quantiles)
+            advisories = (
+                InsufficientQuantiles(
+                    pagination_node,
+                    pagination_field,
+                    current_quantile_resolution,
+                    ideal_quantile_resolution,
+                ),
+            )
+
+    # Construct and return the plan and advisories.
+    if page_capacity <= 1:
+        return None, advisories
+    plan = VertexPartitionPlan((pagination_node,), pagination_field, page_capacity)
+    return plan, advisories
+
+
+def get_pagination_plan(
+    schema_info: QueryPlanningSchemaInfo, query_ast: DocumentNode, number_of_pages: int
+) -> Tuple[PaginationPlan, Tuple[PaginationAdvisory, ...]]:
+    """Make a best-effort PaginationPlan and advise on how to improve statistics.
+
+    Might paginate to fewer than the desired number of pages if no good pagination plan
+    is found. This can happen when there's not enough data, or when the planner is not
+    smart enough to find a good plan. In that case it will return along with the result
+    a list of PaginationAdvisorys that indicate why the pagination was not successful.
+    """
```

Comment:
Since this is more than a one-liner, it needs Args/Returns sections:
```suggestion

    Args:
        schema_info: query planning information, including quantile statistics for pagination
        query_ast: GraphQL AST node describing the query being paginated
        number_of_pages: desired number of pages to attempt to paginate the query into

    Returns:
        tuple including a best-effort pagination plan together with a tuple of advisories describing
        any ways in which the pagination plan was less than ideal and how to resolve them
    """
```